Refactor UPS patching

Nemris opened the pull request:
This PR is meant to refactor the UPS side of ROM patching.

For every commit, I made sure that my test ROM (Mother 3) evaluated to the correct SHA-1 hash after being patched.
Later on, I found a Pokémon romhack that requires resizing of the source ROM, so that part of the logic was tested and confirmed working too. More info in the commit messages.

---

## ROMs tested

Alongside Mother 3 and the related patch, I also tested Pokémon Emerald with the Theta Emerald Renev romhack, which resizes the ROM from 16MiB to 32MiB.

Additionally, I cobbled up a handful of dummy UPS patches in GodMode9 to test edge cases such as a too short patch. The only problem I didn't correct, and which exists in your version, is that certain malformed patches might end up having you force a poweroff by holding Power, which is the main reason I marked the PR as a WIP.

Finally, in order to test whether the empty space padding works as expected for ROMs that are bigger after being patched, I manually trimmed the Emerald Renev romhack to exclude the padding data and adjusted the expected patched ROM size.
The patched OAF was able to pad the ROM and produce the correct hash, but the one from the master branch did not.

---

## Some execution metrics

Thanks to a new rudimentary timing function (which is useless if the patching would exceed 59s), I was able to confirm that the refactor marginally improved patching times.

This is more apparent the bigger the patch is (Theta Emerald Renev takes 21-23s with your debug build but 16-17s on mine) and less obvious for small patches (Mother 3 clocks in at 1-2s in yours and 1s in mine).
The timing logic is compiled out in release mode, but hacky wall-clock measurements show a 3s-ish improvement for Theta Emerald Renev.

---

## Footnotes on the UPS format

A UPS file is binary data which begins with the `UPS1` magic.

Then, two variable-length potentially infinite integers (varints) follow, and represent the source ROM size and the patched ROM size, respectively.

Afterwards there is a list of blocks, starting with a varint that contains the offset to add to our current position in the ROM and followed by the bytes to XOR with our base ROM data. The end of a block is marked with a null byte.

Finally, 12 bytes before the end of the patch, there are three CRC-32 hashes in the ZLIB CRC-32 flavor. The first is the hash of the base ROM, the second is that of the patched ROM and the last is the hash of the patch, minus the last four bytes. We are not using these at the moment, but could do so if OAF had the required CRC-32 implementation.

Closes #247.

Nemris committed:
Exit early if the patched ROM is too big.

Since nextPow2() will produce a value greater than patchedRomSize only
if the latter isn't a power of 2, checking if the size exceeds 32MiB
allows us to exit earlier.

Nemris committed:
Simplify magic checking.

Reduce the horizontal indentation by getting rid of an otherwise
unneeded boolean and guarantee patches will always be valid by the end
of these checks.

Nemris committed:
Compute correct patch length only once.

Provide some validation against too short patch files and precompute
patch size sans CRC-32 hashes. Give a name to some magic values.

Nemris committed:
Conform to the existing code style.

Nemris committed:
Further simplify magic checking.

Since we already prime the cache with the beginning of the patch data,
simply compare the buffer to the expected magic and increment the cache
offset afterwards.

Nemris committed:
Reduce redundancy in Cache field names.

Improve readability since the fields' role is clear from the context.

Nemris committed:
Split UPS metadata loading and verification.

Define a struct holding some basic UPS patch metadata to pass around in
bulk at a later time. Centralize verification of patch invariants
outside the patchUPS() function.

Nemris committed:
Avoid computing patch size more than once.

Pass around the UPSPatch to avoid calling fSize() more than once. Drop
some parentheses to conform with the existing OAF style.

Nemris committed:
Use correct ROM size when patching.

Ensure that the ROM size being used is correct even in case we upscaled
romSize before, since patch.patchedRomSize is the effective size of the
data, but romSize would be the virtual cart capacity.

Nemris committed:
Query cache to know when to stop patching.

Expose a function that shows whether the cache has unread data left.
Stop querying patch file information outside of readCache() for the
purpose of determining how much data is left to read.

Nemris committed:
Move cache init closer to where it's used.

Nemris committed:
Limit varint decoding to u32.

Since valid patches don't produce ROMs exceeding 2^25 bytes, we have no
need for uintmax_t, which is 8 bytes wide.

Nemris committed:
Time UPS patching in debug mode.

The measurement function is rudimentary and only good for times <60s.

Nemris committed:
Extract cache loading.

Hide calls to fRead() within loadCache(). Maintain our own totalRead in
lieu of calling fTell().

Nemris committed:
Make buffer capacity a common constant.

Nemris committed:
Touch ROM bytes only once when upscaling.

Instead of filling the upscaled ROM to capacity with 0xFF and then
zeroing the used region, touch the bytes only once by zeroing out the
used area first and then padding the effectively unused area.

Since some romhacks contain data to manage ROM padding on their own,
this means that the padding memset() might write zero bytes while the
patching process handles the padding on its own.

profi200 commented:
First of all sorry for the slow responses. Currently there is more important stuff for me to invest time into which is why there is barely any activity.

I will review this PR.

Nemris commented:
No problem at all, take all the time you need.

Nemris committed:
Simplify cache loading and ROM patching.

For the patching, take advantage of the fact that `value ^ 0 == value`
in order to streamline the logic.

Nemris commented:
I don't think I can make the patching any faster unless I manage to sidestep even more IPC calls.
That probably means using a larger buffer at this point in order to save on the `fRead()` calls, and indeed I noticed a speed increase when testing with 1KiB and 2KiB buffers.

Any ideas about how much memory is free by the time we reach the patching routine with a 32MiB ROM loaded in RAM?

Edit: 4KiB of buffer space, which is 4x what stock OAF uses, are enough to drive patching times from 20s to 6s for Emerald Renev. Past that point there are no meaningful improvements.

Nemris committed:
Raise buffer size to improve patching speed.

Quadruple the buffer size following testing with Emerald Renev. With
buffers of size 512B to 2KiB, patching takes 20s and 8s, respectively.
Beyond a 4KiB buffer, with which patching takes 6s, there's no
improvement. The approx time elapsed in release mode is 5.5s.

profi200 commented:
We have about 300-400 KiB of RAM left not counting VRAM. That being said if 4 KiB is enough i that's fine.

Nemris commented:
Indeed, returns are virtually nonexistent past 4KiB anyway. Maybe you can see ways to optimize the patching process - but the timings are very close to what I'm getting with a Rust tool I doodled for the occasion and which runs on Android, so who knows.

profi200 commented:
Is there still something you want to improve? Otherwise i will take another look and merge this.

Nemris commented:
There's still the bit about specially-crafted broken UPS patches freezing the patching process, but one shouldn't really run into them in the wild. I think there's nothing more to add to this PR on my side.

(Will probably refactor `patchRom()` in a future, but that's for another time.)

profi200 commented:
I'm going to squash and merge this now otherwise i will never get to it. I'm going to see if i can improve this later/in the future.
